### ISSUE TITLE
Out of CUDA on 15 GB colab .  Just trying to train Mistral 7. v1 

### ISSUE THREAD

**Issue by raymondbernard** [pytorch/torchtune#665]
Here is a a cell I am running

 !tune run recipes/lora_finetune_single_device.py --config recipes/configs/mistral/7B_lora.yaml

by the  way  app has a hard time finding the proger locations of the files in Colab.  

Also I have downloaded Mistral ok 
!tune download mistralai/Mistral-7B-v0.1 --hf-token xxxxxxxxxxx --output-dir /tmp/Mistral-7B-v0.1 

Note == INFO:torchtune.utils.logging:BF16 not supported on this hardware. Setting dtype to float32
The GPU is full before even training .. But Google colab says I have 0 of 15 GB free?? 

Tried the below but failed : 
export PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
import torch
torch.cuda.empty_cache()

Last issues I am using a custom data set alpaca cleaned .. I modified dataset\_alpaca.py .. I konw this is wrong but the docs are not clear : 
    return InstructDataset(
        tokenizer=tokenizer,
        source="RayBernard/vastdata", ## new data source
        template=AlpacaInstructTemplate,
        train_on_input=train_on_input,
        max_seq_len=max_seq_len,
        split="train",
    )

INFO:torchtune.utils.logging:Running recipe_main with parameters {'tokenizer': {'_component_': 'torchtune.models.mistral.mistral_tokenizer', 'path': '/tmp/Mistral-7B-v0.1/tokenizer.model'}, 'dataset': {'_component_': 'torchtune.datasets.alpaca_dataset', 'train_on_input': True}, 'seed': None, 'shuffle': True, 'model': {'_component_': 'torchtune.models.mistral.lora_mistral_7b', 'lora_attn_modules': ['q_proj', 'k_proj', 'v_proj'], 'apply_lora_to_mlp': True, 'apply_lora_to_output': True, 'lora_rank': 64, 'lora_alpha': 16}, 'checkpointer': {'_component_': 'torchtune.utils.FullModelHFCheckpointer', 'checkpoint_dir': '/tmp/Mistral-7B-v0.1', 'checkpoint_files': ['pytorch_model-00001-of-00002.bin', 'pytorch_model-00002-of-00002.bin'], 'recipe_checkpoint': None, 'output_dir': '/tmp/Mistral-7B-v0.1', 'model_type': 'MISTRAL'}, 'resume_from_checkpoint': False, 'optimizer': {'_component_': 'torch.optim.AdamW', 'lr': 2e-05}, 'lr_scheduler': {'_component_': 'torchtune.modules.get_cosine_schedule_with_warmup', 'num_warmup_steps': 100}, 'loss': {'_component_': 'torch.nn.CrossEntropyLoss'}, 'batch_size': 4, 'epochs': 3, 'max_steps_per_epoch': None, 'gradient_accumulation_steps': 1, 'device': 'cuda', 'enable_activation_checkpointing': True, 'dtype': 'bf16', 'metric_logger': {'_component_': 'torchtune.utils.metric_logging.DiskLogger', 'log_dir': '${output_dir}'}, 'output_dir': '/tmp/Mistral-7B-v0.1', 'log_every_n_steps': None}
INFO:torchtune.utils.logging:BF16 not supported on this hardware. Setting dtype to float32
DEBUG:torchtune.utils.logging:Setting manual seed to local seed 3636775082. Local seed is seed + rank = 3636775082 + 0
Writing logs to /tmp/Mistral-7B-v0.1/log_1712616759.txt
Traceback (most recent call last):
  File "/usr/local/bin/tune", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.10/dist-packages/torchtune/_cli/tune.py", line 49, in main
    parser.run(args)
  File "/usr/local/lib/python3.10/dist-packages/torchtune/_cli/tune.py", line 43, in run
    args.func(args)
  File "/usr/local/lib/python3.10/dist-packages/torchtune/_cli/run.py", line 180, in _run_cmd
    self._run_single_device(args)
  File "/usr/local/lib/python3.10/dist-packages/torchtune/_cli/run.py", line 94, in _run_single_device
    runpy.run_path(str(args.recipe), run_name="__main__")
  File "/usr/lib/python3.10/runpy.py", line 289, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/usr/lib/python3.10/runpy.py", line 96, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "recipes/lora_finetune_single_device.py", line 463, in <module>
    sys.exit(recipe_main())
  File "/usr/local/lib/python3.10/dist-packages/torchtune/config/_parse.py", line 54, in wrapper
    sys.exit(recipe_main(conf))
  File "recipes/lora_finetune_single_device.py", line 457, in recipe_main
    recipe.setup(cfg=cfg)
  File "recipes/lora_finetune_single_device.py", line 152, in setup
    self._model = self._setup_model(
  File "recipes/lora_finetune_single_device.py", line 221, in _setup_model
    model = config.instantiate(cfg_model)
  File "/usr/local/lib/python3.10/dist-packages/torchtune/config/_instantiate.py", line 103, in instantiate
    return _instantiate_node(config, *args)
  File "/usr/local/lib/python3.10/dist-packages/torchtune/config/_instantiate.py", line 31, in _instantiate_node
    return _create_component(_component_, args, kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torchtune/config/_instantiate.py", line 20, in _create_component
    return _component_(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torchtune/models/mistral/_model_builders.py", line 77, in lora_mistral_7b
    return lora_mistral(
  File "/usr/local/lib/python3.10/dist-packages/torchtune/models/mistral/_component_builders.py", line 219, in lora_mistral
    model = TransformerDecoder(
  File "/usr/local/lib/python3.10/dist-packages/torchtune/modules/transformer.py", line 137, in __init__
    self.layers = _get_clones(layer, num_layers)
  File "/usr/local/lib/python3.10/dist-packages/torchtune/modules/transformer.py", line 93, in _get_clones
    return nn.ModuleList([copy.deepcopy(module) for i in range(n)])
  File "/usr/local/lib/python3.10/dist-packages/torchtune/modules/transformer.py", line 93, in <listcomp>
    return nn.ModuleList([copy.deepcopy(module) for i in range(n)])
  File "/usr/lib/python3.10/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.10/copy.py", line 271, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python3.10/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.10/copy.py", line 231, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.10/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.10/copy.py", line 297, in _reconstruct
    value = deepcopy(value, memo)
  File "/usr/lib/python3.10/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.10/copy.py", line 271, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python3.10/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.10/copy.py", line 231, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.10/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.10/copy.py", line 297, in _reconstruct
    value = deepcopy(value, memo)
  File "/usr/lib/python3.10/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.10/copy.py", line 271, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python3.10/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.10/copy.py", line 231, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.10/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.10/copy.py", line 297, in _reconstruct
    value = deepcopy(value, memo)
  File "/usr/lib/python3.10/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.10/copy.py", line 271, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python3.10/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.10/copy.py", line 231, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.10/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.10/copy.py", line 297, in _reconstruct
    value = deepcopy(value, memo)
  File "/usr/lib/python3.10/copy.py", line 153, in deepcopy
    y = copier(memo)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/parameter.py", line 59, in __deepcopy__
    result = type(self)(self.data.clone(memory_format=torch.preserve_format), self.requires_grad)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_device.py", line 77, in __torch_function__
    return func(*args, **kwargs)
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 2.00 MiB. GPU 0 has a total capacity of 14.75 GiB of which 3.06 MiB is free. Process 184044 has 14.74 GiB memory in use. Of the allocated memory 14.60 GiB is allocated by PyTorch, and 29.15 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

**Comment by RdoubleA** [comment]
Hey @raymondbernard thanks for posting this issue, as I'm sure many users will be fine-tuning on Colab!

The reason for the OOM is primarily this log:

>INFO:torchtune.utils.logging:BF16 not supported on this hardware. Setting dtype to float32

I'm assuming you are using the free version of Colab which uses T4s. These unfortunately do not support bf16 precision, which means you will need double the memory just to load a 7B model (~28 GB). I would recommend trying to upgrade to a GPU that does support bf16.

We are working on adding smaller models (< 7B) that could potentially be fine-tuned on Colab, so please look out for that!

**Comment by raymondbernard** [comment]
Thank you for your response!

**Comment by ebsmothers** [comment]
Hey @raymondbernard, thanks for the issue! One other comment here: as @RdoubleA said, we typically load weights in either bf16 or fp32, so since T4 does not support bf16 we load in fp32. But there are other ways that weights can be loaded. E.g. with QLoRA we load many of the weights in 4 bits, so even if the remaining weights are in fp32 you should still be able to load the full model in under 16GB. There is one thing blocking a proper finetune on Colab, which is how we validate that the weights are loaded correctly. I am opening #666 to address this, but after that you should at least be able to run a fine-tune of Mistral 7B with QLoRA on the Colab T4s.

**Comment by rohan-varma** [comment]
@raymondbernard Thanks for filing this issue! We recently enabled QLoRA in fp32: https://github.com/pytorch/torchtune/pull/595, so after @ebsmothers unblocking work on validate_state_dict, this finetune should be enabled.

Though @ebsmothers, if the bottleneck of `validate_state_dict` causes OOM on T4, then when we take actual `state_dict`, won't we have similar issue of the memory peaking?

**Comment by ebsmothers** [comment]
@rohan-varma yeah thanks for pointing this out. As is we will OOM eventually on checkpoint save regardless (verified this myself in a T4 colab notebook). I don't know offhand what the best way around this is, I think we would need some follow-up work here to properly support checkpoint saving in Colab. @raymondbernard are you hoping to use Colab to save checkpoints as well, or is the idea here to just try out the model and train a few iterations? It'd be helpful to know more about your intended usage here.

**Comment by raymondbernard** [comment]
@ebsmothers the later point.  "the idea here to just try out the model and train a few iterations"

**Comment by razi-tm** [comment]
hi, thanks for providing open source tools. I'm getting cuda oom error when fine-tuning llama 3 8b on a 12gb gpu with bf16 and qlora config. shouldn't its peak memory usage be 9gb?

**Comment by ebsmothers** [comment]
@raymondbernard sorry for the delayed response here. We made some changes since the issue opened, and you should now be able to train with QLoRA in fp32 on Colab T4. However you still cannot save checkpoints (since the weights are saved in fp32), though we are working on that as well.

@razi-tm what type of GPU are you using? Does it support bf16 dtype?

**Comment by razi-tm** [comment]
rtx 3060 and it supports bf16

**Comment by ebsmothers** [comment]
@razi-tm thanks for flagging this, I am looking into it now. Out of curiosity, are you able to start training before seeing the OOM or do you see it during checkpoint load?

**Comment by razi-tm** [comment]
it happens at the begging of training

**Comment by razi-tm** [comment]
it happens at the beginning of training

**Comment by RdoubleA** [comment]
original issue was addressed, closing now. Please reopen a new issue if you continue to see problems with QLoRA finetuning.
